PetrolCar.cpp

```diff
@@ -1,16 +1,67 @@
-#include "PetrolCar.hpp"
 #include <iostream>
+
+#include "PetrolCar.hpp"
  
 PetrolCar::PetrolCar(PetrolEngine* engine)
     : engine_(engine)
 {
-    std::cout << __FUNCTION__ << std::endl;
+    std::cout << "* " << __FUNCTION__ << std::endl;
+}
+
+PetrolCar::~PetrolCar() {
+    std::cout << "* " << __FUNCTION__ << std::endl;
+    delete engine_;
+}
+
+void PetrolCar::refuel() {
+    std::cout << __FUNCTION__;
+    if (fuelLevel_ < MAX_CAPACITY) {
+        fuelLevel_ = MAX_CAPACITY;
+        std::cout << " : refuelled" << std::endl;
+    } else {
+        std::cerr << " : tank already full" << std::endl;
+    }
 }
 
-PetrolCar::~PetrolCar()         { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
-void PetrolCar::refuel()        { std::cout << __FUNCTION__ << std::endl; }
+size_t PetrolCar::getFuelLevel() {
+    return fuelLevel_;
+}
+
+void PetrolCar::setEngine(PetrolEngine* engine) {
+    std::cout << __FUNCTION__;
+    if (!getSpeed()) {
+        std::cout << " : engine changed" << std::endl;
+        delete engine_;
+        engine_ = engine;
+        std::cout << "@ petrol power: " << engine_->getPower() << std::endl;
+    } else {
+        delete engine;
+        std::cout << " : not possible unless stopped" << std::endl;
+    }
+}
+void PetrolCar::setEngine(ElectricEngine*) {}
+void PetrolCar::setEngine(PetrolEngine*, ElectricEngine*) {}
+
 
+void PetrolCar::restore() {
+    refuel();
+}
+
+void PetrolCar::changeGear(int gear) {
+    std::cout << __FUNCTION__;
+    if (gear >= -1 && gear <= engine_->getGears()) {
+        // -1 is for REAR
```

Comment:
Hi nice that you explain meaning of -1. However I think that is better to use constexpr or const variable.  What do you think about thrownig an exception: InvalidGear for some cases?

Reply:
It was so in the original code, so I left it. And indeed it is better to avoid such numbers and use constexpr instead. Throwing exception is fine - here standard output was used to verbose the problems and not to stop the program. Thanks for your comment Maciej!

Reply:
Constexpr for reverse gear and exception handling added, finally!